### DISCUSSION TITLE
Add sensor device class for reactive energy

### DISCUSSION THREAD

**Discussion by toreamun** [home-assistant/architecture#724] [Entity Models] [answered]
Support for device class reactive power (var) has just been added by #688 and https://github.com/home-assistant/core/pull/62010. Reactive energy (varh/kvarh) is often also used when reactive power is used. Reactive energy should be added as a device class and varh and kvarh as units.

**Comment by emontnemery** [comment]
Can you try to list integrations which could make use of this?

Also, maybe a bit off topic, but do you know cases where consumers will be charged for reactive energy?

**Reply by farmio** [reply]
`Fronius` would support it. 👍

**Reply by toreamun** [reply]
At least Fronius and IoTaWatt. Maybe Tibber. My custom integration https://github.com/toreamun/amshan-homeassistant for Norwegian and Swedish also need this. 

The national standard of Norway (https://www.nek.no/info-ams-han-utviklere/) and Sweden (https://www.energiforetagen.se/globalassets/energiforetagen/det-erbjuder-vi/kurser-och-konferenser/elnat/branschrekommendation-lokalt-granssnitt-v2_0-201912.pdf) includes reactive energy and power.

I don't know of any cases where consumers will be charged for reactive energy. But reactive energy is measured by ALL meters in Norway.

**Reply by DennisHedberg** [reply]
There are many versions of custom ESP32 solutions that talks with the Powermeter via HAN / P1 port. Integration done via standard MQTT with HA.
As stated above some of the sensors needs a device class for Reactive Energy, varh, kvarh 
All new meters in Sweden have this.

I think only factories, business etc. pay for Reactive Energy in Sweden today.

https://energynordic.com/sv/reactive-energy-impact-sme-electricity-bill/

**Reply by balazs92117** [reply]
Both kvar and kvarh units are provided via P1 port (DSMR) in Hungary too!

**Reply by balazs92117** [reply]
https://www.home-assistant.io/integrations/dsmr/ could use reactive energy (varh/kvarh) too.

**Reply by banditdk** [reply]
All electricity power meters have this unit, should be supported in HA.
SensorDeviceClass.APPARENT_POWER is allready supported, SensorDeviceClass.REACTIVE_POWER is not.

**Reply by Breina** [reply]
We can add my integration to this list as well. It's about energy monitoring per circuit.

https://github.com/Breina/PowerTagGateway

**Reply by NECH2004** [reply]
In Austria all new smart meters using MBUS provide reactive energy (and also power). It is part of the OBIS standard.
1-0:3.8.0.255 - Reactive energy consumed +R  (varh)
1-0:4.8.0.255 - Reactive energy returned -R (varh)

As in Poland you have to pay for reactive energy if you are a large consumer or also producer.

**Comment by piotrtobolski** [comment]
As a result of discussion #688 support was added for reactive and apparent power. Energy meters that support those values report also reactive and apparent energy. Basically all integrations that support reactive and apparent power would benefit from this. It seems like a no-brainer to add this class to HA with respective units.

There were already bug and even PRs on core to add this as it is quite simple change but both were rejected until architectural decision is made:
https://github.com/home-assistant/core/issues/85507
https://github.com/home-assistant/core/pull/98453

Can we have this decision made? It seems like a simple one.

cc @emontnemery @MartinHjelmare

**Reply by MartinHjelmare** [reply]
Someone should do a thorough analysis of existing integrations and list all that can use it. So far only 3-4 integrations have been mentioned which seems on the low side for a new device class.

**Reply by piotrtobolski** [reply]
If there was enough reason to add reactive and apparent power, there is also reason for reactive and apparent energy if there is already active energy implemented. Those classes are complementary. It is weird to include reactive, apparent and active power but active energy. I still think it looks like a bug that someone forgot to add those classes when adding new power types, not like something that should require architectural decision.

**Reply by banditdk** [reply]
it is fairly hard to mention integrations, when it is a standard data point in all power meters. And most are receiving this data from the power meter via esp home or mqtt. Not a bunch of individual integrations.
Abit like only having kW but not supporting kWH.

**Reply by parautenbach** [reply]
Out of honest interest: What would the general user do with this information? Many users don't even understand the difference between power and energy.

**Reply by piotrtobolski** [reply]
In some countries you pay not only for active energy but also for reactive one. AFAIK it will be more and more popular due to high reactive power draw from the heat pumps. In Poland for now only big businesses pay for the reactive power but in some cases even households may need to pay for it.

Reporting apparent or reactive power is standard feature on almost all power meters. I use modbus tcp to integrate with my meter but there are plenty other ways to connect to a power meter.

**Reply by dgomes** [reply]
I bet many use-cases involve MQTT, hard to count those.

**Reply by Breina** [reply]
@parautenbach My project is used by electricity professionals as well, so I guess this information is relevant for tracking a circuit's performance in industrial settings.

**Reply by NECH2004** [reply]
I've looked into the source code.
Fronius, Tasmota and iotawatt are using varh or kvarh (sometimes different spellings).
Maybe also dsmr will need it in future (because of the OBIS standard).
@Breina has the need (and also my custom integration)

**Reply by NECH2004** [reply]
Here is an other link that asks for a reactive energy sensor - but was closed because of two proposals.
https://github.com/home-assistant/architecture/discussions/955

**Reply by AlCalzone** [reply]
I'll add Z-Wave to the list of integrations that can make use of it.
Related: https://github.com/zwave-js/zwave-js-ui/issues/3956, https://github.com/home-assistant/core/issues/128325

**Reply by puddly** [reply]
Zigbee uses BACnet units for some sensors, so ZHA can be added to the list.

**Comment by Passific** [comment]
I was trying to [add](https://github.com/home-assistant/core/pull/143941) support for this unit also but got redirected here.
So after 3 years, can we expect to have an approval (or denial, which would be surprising)?

**Reply by abmantis** [reply]
I guess this that this went stale because the original discussion post was never updated to address [this request](https://github.com/home-assistant/architecture/discussions/724#discussioncomment-2102735): 

> Can you try to list integrations which could make use of this?

Also https://github.com/home-assistant/architecture/discussions/724#discussioncomment-9294099


Usually new device classes are only added if they can be used by multiple existing official integrations (not custom ones).

**Reply by Passific** [reply]
> multiple existing official integrations (not custom ones).

The custom integration I'm talking in my PR is about the serial link with the utility meter, but there is also a Zigbee version, which is supported by [ZHA](https://github.com/zigpy/zha-device-handlers/issues/1146), and could benefits from this.

**Comment by piotrtobolski** [comment]
It looks like we need to create list of integrations that would benefit from adding this new device class. Can we do it in comments below?

**Reply by piotrtobolski** [reply]
[Modbus](https://www.home-assistant.io/integrations/modbus/) - I use it together with Modbus TCP adapter to integrate energy meter
[MQTT Sensor](https://www.home-assistant.io/integrations/sensor.mqtt/) - In case energy meter is integrated through MQTT
[Template](https://www.home-assistant.io/integrations/template/) - Obvious
[Tasmota](https://www.home-assistant.io/integrations/tasmota/) - Many energy plugs running Tasmota allow measuring reactive power and energy
[Z-Wave](https://www.home-assistant.io/integrations/zwave_js/) - In case energy meter is integrated through Z-Wave
[Zigbee Home Automation](https://www.home-assistant.io/integrations/zha/) - In case energy meter is integrated through Zigbee
[Fronius](https://www.home-assistant.io/integrations/fronius/) - Integrating solar inverters

**Reply by piotrtobolski** [reply]
[DSMR Smart Meter](https://www.home-assistant.io/integrations/dsmr/) - Smart meter integration

**Reply by BradleyFord** [reply]
via BACNET this can also be retrieved. 
as was previously mentioned nearly all energy meters support this. 
(by law in Thailand all solar systems need a Modbus energy meters, of which they all support reactive)

**Reply by Breina** [reply]
[Schneider PowerTags](https://github.com/Breina/PowerTagGateway) - measure electricity characteristics on a per-circuit level.

Thanks for swinging this back into action!

**Comment by Hedda** [comment]
FYI, I believe that correct terminology in the English language is either ”Volt-ampere reactive”, hence ”var” or ”VAR” as abriviation, (or just ”Volt-ampere” for short). 

Also note that that should add Volt-ampere reactive per hour ”varh” or ”VARh” for tracking energy usage.

While the terms Reactive Power, Reactive Energy, or Line Voltage are more for everyday use by laymans to describe the measurement of reactive power.

From wikipedia:

The unit "var" is allowed by the  [International System of Units](https://en.m.wikipedia.org/wiki/International_System_of_Units) (SI) even though the unit var is representative of a form of power. Per EU [directive 80/181/EEC](https://en.m.wikipedia.org/wiki/Directive_80/181/EEC) (the "metric directive"), the correct symbol is lower-case "var", although the spellings "Var" and "VAr" are commonly seen, and "VAR" is widely used throughout the power industry.

**Reply by parautenbach** [reply]
Pedantic note: VARh is VA x h not VA/h.

**Comment by abmantis** [comment, marked as answer]
This proposal was discussed in today's architecture core meeting and was accepted. PRs implementing this can now be open :+1:

**Reply by frenck** [reply]
This change has been merged/added.

**Comment by Breina** [comment]
Shit we forgot about apparent energy.
